packages/driver/cypress/e2e/commands/assertions.cy.js

```diff
@@ -302,6 +302,19 @@ describe('src/cy/commands/assertions', () => {
         })
       })
 
+      // https://github.com/cypress-io/cypress/issues/22587
+      it('does not allow cypress commands inside the callback', (done) => {
```

Comment:
This may be a silly question, but if we're being more prescriptive about the use of `should`, should we allow _any_ callbacks with `should`? 

Note: There's probably/definitely use cases I'm not aware of, but I thought `should` was only meant to be used with assertions like `should('exist')` rather than with callbacks

Reply:
Ah I see the [docs do show you can use callbacks](https://docs.cypress.io/api/commands/should#Chainers), but would it be better not to? 🤷‍♀️

Reply:
https://docs.cypress.io/api/commands/should#Function has a lot of examples of using a function, and I think "accepting a user-supplied callback function" has fairly wide applicability.

Now that I look at the docs (which I probably should have started with!), it looks like we already have guidance for users around the contents of their functions.

> Be sure not to include any code that has side effects in your callback function. The callback function will be retried over and over again until no assertions within it throw.

So this PR is just being a bit stricter about it, saying "all cypress commands have side effects (on the cypress command queue and the command log, if nothing else)."

Reply:
I like your change, helping our users not get stuck in these scenarios in the first place